rlcard/games/gin_rummy/card.py

```diff
@@ -4,8 +4,6 @@
     Date created: 2/12/2020
 '''
 
-from typing import List
-
 
 class Card(object):
```

Comment:
Can we reuse the Card defined in https://github.com/datamllab/rlcard/blob/master/rlcard/core.py#L4

Reply:
I need my Card class to have a hash function that treats cards with same suit and rank as being equivalent. I am sometimes putting my cards in a Python set. Hashing also requires equality to be defined. 

I can add the following to the Card class:

   
    def __eq__(self, other):
        if isinstance(other, Card):
            return self.rank == other.rank and self.suit == other.suit
        else:
            # don't attempt to compare against unrelated types
            return NotImplemented

    def __hash__(self):
        suit_index = Card.valid_suit.index(self.suit)
        rank_index = Card.valid_rank.index(self.rank)
        return rank_index + 100 * suit_index

Is this ok? If you want, I could change "return NotImplemented" with "return False"

Reply:
Looks good!

Reply:
Can I also add an implementation for _ _str_ _ ?
 
   
    def __str__(self):
        ''' Get string representation of a card.

        Returns:
            string: the combination of rank and suit of a card. Eg: AS, 5H, JD, 3C, ...
        '''
        return self.rank + self.suit

This is useful in debugging since this string representation is used instead of just showing the information like: <rlcard.core.Card object at 0x10e49d3c8>. If _ _str_ _ is implemented than the debugger shows the card name like: "KS".

It is also useful when doing print statements since one does not need to convert "self.card" to its string representation since it is done automatically in a print statement. That is, e.g., instead of writing:

    return "discard {}".format(str_card(self.card))

I can write:

    return "discard {}".format(self.card)

According to the web:

=======================================

This method returns the string representation of the object. This method is called when  print() or str() function is invoked on an object.

This method must return the String object. If we don’t implement _ _str_ _() function for a class, then built-in object implementation is used that actually calls _ _repr_ _() function.

=======================================

Reply:
It seems that the function is the same as the `get_index`. The difference is only the order of suit and rank. Do you think it would be better to rename `get_index` to `__str__`? Another thing is the order of self.rank and self.suit. It seems that the `get_index` uses suit + rank while your `__str__` uses rank+suit. Which representation do you plan to use in your game engine?

Reply:
The Python  _ _ str _ _ function interacts with Python's print and str functions while get_index does not. Also, the _ _ str _ _ function interacts with most debuggers to give a readable output of a variable rather than the <rlcard.core.Card object at 0x10e49d3c8> format. The get_index method does not do this (since the debugger doesn't know about get_index).

Again, I am not sure what you mean by which representation do I plan to use in the game engine. I am already using the get_index representation in my "state" as was requested. But, I am going to use the "KS" format for my printout. And in my machine representation I am just using a 1/0 in a 52-dim array. I think the "SK" format is really just a typo. Look at the documentation for the get_index method: it does say suit followed by rank, but then it gives examples of "1S" (a typo, should be "AS"), "2S", etc.

Reply:
Thanks for the clarification. We can use this `__str__`